query-engine/core/src/interactive_transactions/manager.rs

```diff
@@ -0,0 +1,192 @@
+use crate::{ClosedTransaction, InteractiveTransaction, Operation, ResponseData};
+use connector::Connection;
+use lru::LruCache;
+use once_cell::sync::Lazy;
+use schema::QuerySchemaRef;
+use std::{collections::HashMap, sync::Arc};
+use telemetry::helpers::TraceParent;
+use tokio::{
+    sync::{
+        mpsc::{unbounded_channel, UnboundedSender},
+        Mutex, RwLock,
+    },
+    time::Duration,
+};
+use tracing_futures::WithSubscriber;
+
+#[cfg(not(feature = "metrics"))]
+use crate::metrics::MetricsInstrumentationStub;
+#[cfg(feature = "metrics")]
+use prisma_metrics::WithMetricsInstrumentation;
+
+use super::{TransactionError, TxId};
+
+pub static CLOSED_TX_CACHE_SIZE: Lazy<usize> = Lazy::new(|| match std::env::var("CLOSED_TX_CACHE_SIZE") {
+    Ok(size) => size.parse().unwrap_or(100),
+    Err(_) => 100,
+});
+
+pub struct ItxManager {
+    /// Stores all current transactions (some of them might be already committed/expired/rolled back).
+    ///
+    /// There are two tiers of locks here:
+    ///  1. Lock on the entire hashmap. This *must* be taken only for short periods of time - for
+    ///     example to insert/delete transaction or to clone transaction inside.
+    ///  2. Lock on the individual transactions. This one can be taken for prolonged periods of time - for
+    ///     example to perform an I/O operation.
+    ///
+    /// The rationale behind this design is to make shared path (lock on the entire hashmap) as free
+    /// from contention as possible. Individual transactions are not capable of concurrency, so
+    /// taking a lock on them to serialise operations is acceptable.
+    ///
+    /// Note that since we clone transaction from the shared hashmap to perform operations on it, it
+    /// is possible to end up in a situation where we cloned the transaction, but it was then
+    /// immediately removed by the background task from the common hashmap. In this case, either
+    /// our operation will be first or the background cleanup task will be first. Both cases are
+    /// an acceptable outcome.
+    transactions: Arc<RwLock<HashMap<TxId, Arc<Mutex<InteractiveTransaction>>>>>,
```

Comment:
I was thinking about using `DashMap` here btw but I'm not sure if it's worth adding a new dependency. As Nikita wrote above, even though we use coarse-grained locking on the hash map, the lock is only held for a tiny fraction of time. I'm actually thinking even `RwLock` might not be worth it here and a regular `Mutex` may be just fine while having lower overhead. All of this is moot without benchmarks of course, just thinking out loud.